### DISCUSSION TITLE
Proposal: Generics Type Arguments of class recognized by class constructor

### DISCUSSION THREAD

**Discussion by pmunin** [dotnet/csharplang#427] [General]
It'd be really nice if class constructors could recognize Generic Type Argument for the class. That would allow to instantiate generic classes for Anonymous types much easier without helper methods.

**Actual behavior** :-1: :-1: :-1:
```
var myObjectsArray = new[]{ new{ FirstName: "John", LastName:"Smith"} };
var myObjsQueue = ToQueue(items);

Queue<T> ToQueue<T>(this IEnumerable<T> items)
{
      return new Queue<T>(items);
}
```
**Desired behavior** :+1: :+1: :+1:
```
var myObjsQueue = new Queue<>(new []{ new {FirstName:"John", LastName:"Smith"} });
```

**Another examples of Desired behavior where it could be sweet** :candy: :cake: :honey_pot:
```
var myLazyObj = new Lazy<>(()=>{ 
        var data = DownloadAndCacheData().Wait();
        return new {FirstName:data["firstName"], LastName:data["LastName"]};
}

var myTupleObj = new Tuple<,>("myLazyObj", myLazyObj);

var myDictionary = new Dictionary<,>(){
   {"person1" , new {FirstName:"John", LastName:"Smith"} }
};
```

**Additional nice-to-haves** :star: :star: :star:

- **Smart As/Is operator recognition**:
```
var myDictionary = new Dictionary<,>(){
   {"person1" , new {FirstName:"John", LastName:"Smith"} }
};
var readOnlyDictionary = myDictionary as IReadOnlyDictionary<,>;
```
_Potential issue:_ class can implement multiple interfaces with different generic types e.g.:
MyCustomDictionary: IDictionary<string, T> and IDictionary<object,T>

_Potential solution:_ for that situation return compilation error, something like "Cannot do inference of generic type arguments, because unambiguous...".

- **Local Type alias for anonymous type or "guessed" generic type arguments**
Created separate issue #430, but specifics for this issue could be "guessed" generic type argument:
```
var myDictionary = new Dictionary< ,<TPerson> >(){
   {"person1" , new {FirstName:"John", LastName:"Smith"} }
};

var persons = new List<TPerson>(myDictionary.Values);
```

**Comment by svick** [comment]
This looks like a duplicate of https://github.com/dotnet/csharplang/issues/151.

**Comment by pmunin** [comment]
It does look similar, however I believe definition in this issue is more clean and straightforward. Am I bias? :smile:
